daemon/daemon_main.go

```diff
@@ -1454,3 +1415,130 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
+			option.EnableExternalIPs, option.EnableHostReachableServices)
+
+		option.Config.EnableNodePort = false
+		option.Config.EnableExternalIPs = false
+		option.Config.EnableHostReachableServices = false
+		option.Config.EnableHostServicesTCP = false
+		option.Config.EnableHostServicesUDP = false
+		option.Config.DisableK8sServices = true
+
+		return
+	}
+
+	// strict denotes to panic if any to-be enabled feature cannot be enabled
+	strict := option.Config.KubeProxyReplacement != option.KubeProxyReplacementProbe
+
+	if option.Config.KubeProxyReplacement == option.KubeProxyReplacementProbe ||
+		option.Config.KubeProxyReplacement == option.KubeProxyReplacementStrict {
+
+		log.Infof("Auto-enabling %q, %q, %q features", option.EnableNodePort,
+			option.EnableExternalIPs, option.EnableHostReachableServices)
+
+		option.Config.EnableNodePort = true
+		option.Config.EnableExternalIPs = true
+		option.Config.EnableHostReachableServices = true
+		option.Config.EnableHostServicesTCP = true
+		option.Config.EnableHostServicesUDP = true
+		option.Config.DisableK8sServices = false
+	}
+
+	if option.Config.EnableNodePort {
+		if option.Config.EnableIPSec {
+			msg := "IPSec cannot be used with NodePort BPF."
+			if strict {
+				log.Fatal(msg)
+			} else {
+				option.Config.EnableNodePort = false
+				option.Config.EnableExternalIPs = false
+				log.Warn(msg + " Disabling the feature.")
+			}
+		}
+
+		if option.Config.NodePortMode != "dsr" && option.Config.NodePortMode != "snat" {
+			log.Fatalf("Invalid value for --%s: %s", option.NodePortMode, option.Config.NodePortMode)
+		}
+
+		if option.Config.NodePortMode == "dsr" && option.Config.Tunnel != option.TunnelDisabled {
+			// Currently, DSR does not work in the tunnel mode. Once it's fixed,
+			// the constraint can be removed.
+			log.Fatal("DSR cannot be used with tunnel")
+		}
+	}
+
+	if option.Config.EnableNodePort {
+		found := false
+		if h := probes.NewProbeManager().GetHelpers("sched_act"); h != nil {
+			if _, ok := h["bpf_fib_lookup"]; ok {
+				found = true
+			}
+		}
+		if !found {
+			msg := "BPF NodePort services needs kernel 4.17.0 or newer."
+			if strict {
+				log.Fatal(msg)
+			} else {
+				log.Warn(msg + " Disabling the feature.")
+				option.Config.EnableNodePort = false
+				option.Config.EnableExternalIPs = false
+			}
+		}
+	}
+
+	if option.Config.EnableNodePort && option.Config.Device == "undefined" {
+		device, err := linuxdatapath.NodeDeviceNameWithDefaultRoute()
+		if err != nil {
+			msg := "BPF NodePort's external facing device could not be determined. Use --device to specify."
+			if strict {
+				log.WithError(err).Fatal(msg)
+			} else {
+				log.WithError(err).Warn(msg + " Disabling BPF NodePort feature.")
+				option.Config.EnableNodePort = false
+				option.Config.EnableExternalIPs = false
+			}
+		}
+		log.WithField(logfields.Interface, device).
+			Info("Using auto-derived device for BPF node port")
```

Comment:
I think that this message will always be printed even if the feature was disabled due to a non-fatal error a few lines up, in `--probe` mode?

Reply:
Nice spot!